src/app/upgrade.h

```diff
@@ -112,14 +112,28 @@ bool upgradeResumeFile(const QString &filepath, const QVariantHash &oldTorrent =
     fastNew = fastOld;
 
     bool v3_3 = false;
+    bool artifact = false;
     int queuePosition = 0;
     QString outFilePath = filepath;
-    QRegExp rx(QLatin1String("([A-Fa-f0-9]{40})\\.fastresume\\.(\\d+)$"));
+    QRegExp rx(QLatin1String("([A-Fa-f0-9]{40})\\.fastresume\\.(.+)$"));
     if (rx.indexIn(filepath) != -1) {
-        // old v3.3.x format
+        // Old v3.3.x format had a number at the end indicating the queue position.
+        // The naming scheme was '<infohash>.fastresume.<queueposition>'.
+        // However, QSaveFile, which uses QTemporaryFile internally, might leave artifacts
+        // behind after eg a crash. These artifacts have the following naming scheme
+        // '<infohash>.fastresume.XXXXXX' where each X is random character.
+        // So we detect if the last part is present. Then check if it is 6 chars long.
+        // If all the 6 chars are digits we assume it is an old v3.3.x format.
+        // Otherwise it is considered an artifact which is discarded, because it may
+        // be a corrupted/incomplete fastresume.
+        // NOTE: When the upgrade code is removed, we must continue to perform cleanup of
+        // QSaveFile/QTemporaryFile artifacts
         queuePosition = rx.cap(2).toInt();
```

Comment:
I would just check rx.cap(2) with regex instead of misunderstandable expression below.

Reply:
The expression below is explained in the comment above. I like it better this way.

Reply:
No, because:
1. Assume this name `hash.fastresume.000001`
2. `queuePosition` will be `1`. But old v3_3 format would have never written it like that. It would have written it as `hash.fastresume.1`
3. In case it isn't clear I test for 6 chars because a name like `hash.fastresume.13bafr` would result in `queuePosition == 13`

Reply:
>Assume this name hash.fastresume.000001

So what's wrong with my expression? It matches this filename as artifact and the file will be deleted.

Reply:
I think it can be simplified to:
```c++
queuePosition = rx.cap(2).toInt();  // returns 0 when conversion failed
if (!((99999 < queuePosition) && (queuePosition <= 999999))) {
    Utils::Fs::forceRemove(filepath);
    return true;
}
```

Reply:
Damn it. You're right. @glassez I read your number as six nines not as five nines.
@Chocobo1 I think your condition will also delete valid v3_3 files. eg `hash.fastresume.12345`

I am pushing a version with @glassez's suggestion.